### ISSUE TITLE
Chests can be forced to stay open if formspec is replaced

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest_game#2031]
If a player opens chest formspec, but while the chest formspec is open, another mod sends another formspec for some reason. This can happen at any time since mods are completely free to send in formspec whenever they like.

Now the chest formspec is never officially closed and the player is never removed from the chest list. So the chest stays open. The `receive_fields` is NOT called when an already open formspec is replaced with a new one, so it's not easy to detect.

How to test: Place a chest, then with `luacmd`, do:

`/lua minetest.after(3, function() minetest.show_formspec(me:get_player_name(), "test", "field[test;test;test]") end)`

Then immediately open the chest after hitting enter. The chest formspec should be replaced by a test formspec. Hit Esc, and your chest is still open.

**Comment by 4w** [comment]
> The `receive_fields` is NOT called when an already open formspec is replaced with a new one, [...]

So this is not a chests issue but a general formspec bug ...

**Comment by paramat** [comment]
@sofar

**Comment by Wuzzy2** [comment]
@paramat: I edited the issue to remove the bug of “chest stays open when quitting Minetest”.
Now it should be an all-new issue.

**Comment by HybridDog** [comment]
You could change the chest opening behaviour:
* right click on a closed chest: open it
* right click on an open chest: show formspec
* punch an open chest: close it

This would also fix the unrealistic and annoying waiting time when opening chests: Players right click them, wait until it opens and then view the inventory.
Players would click to open, then click to access, in reality you can open chests fast and harshly or in the slow quiet way.

**Comment by HybridDog** [comment]
l've found another bug: https://github.com/minetest/minetest_game/blob/master/mods/default/nodes.lua#L2085
closed and open chest both use the same on_rightclick function, so when a player right clicks an open chest, it's opened, i.e. the sound is played and the node is replaced with itself. Or did l miss something?
#
When the server shuts down or the formspec is replaced, the formspec actually becomes closed.
Doesn't this issue actually represent an engine bug?

**Comment by paramat** [comment]
@sofar

**Comment by ClobberXD** [comment]
This also happens when the right-click interval is ridiculously small, and two adjacent chests are opened.

## Steps to reproduce
- Set Rightclick Repetition Interval to 0.1 or less
- Ensure there are two chests in adjacent nodes
- Point at one chest and hold down RMB
- Immediately point at the adjacent chest and release RMB

Note: The above procedure must be done very quickly - it must be completed before the formspec for the first chest pops up.

## Result
Minetest registers a right-click for both the chests before displaying the formspec for the first chest, plays the chest-opening anim. and sound, but displays the formspec for the second chest. The first chest stays open until right-clicked again, after which it closes as intended.

**Comment by appgurueu** [comment]
Can be worked around by overriding `show_formspec` to call `on_receive_fields` if the last shown formspec has not been closed.
